src/main/java/uk/gov/pay/directdebit/mandate/dao/MandateSearchDao.java

```diff
@@ -117,17 +125,29 @@ private Params(MandateSearchParams mandateSearchParams) {
             this.reference = new SimpleEntry<>(MANDATE_SERVICE_REFERENCE_COLUMN_NAME, 
                     mandateSearchParams.getReference());
             this.mandateState = new SimpleEntry<>(MANDATE_STATE_COLUMN_NAME, 
-                    Optional.ofNullable(mandateSearchParams.getMandateState()).map(m -> m.name().toLowerCase()).orElse(null));
+                    Optional.ofNullable(mandateSearchParams.getMandateState()).map(m -> m.name()).orElse(null));
             this.mandateBankStatementReference = new SimpleEntry<>(MANDATE_REFERENCE_COLUMN_NAME, 
                     Optional.ofNullable(mandateSearchParams.getMandateBankStatementReference()).map(m -> m.toString()).orElse(null));
             this.name = new SimpleEntry<>(PAYER_NAME_COLUMN_NAME, mandateSearchParams.getName());
             this.email = new SimpleEntry<>(PAYER_EMAIL_COLUMN_NAME, mandateSearchParams.getEmail());
-//            this.fromDate = new AbstractMap.SimpleEntry<>("created_date", mandateSearchParams.getFromDate());
-//            this.toDate = new AbstractMap.SimpleEntry<>("created_date", mandateSearchParams.getToDate());
+            this.fromDate = new SimpleEntry<>(MANDATE_CREATED_DATE_COLUMN_NAME, mandateSearchParams.getFromDate());
+            this.toDate = new SimpleEntry<>(MANDATE_CREATED_DATE_COLUMN_NAME, mandateSearchParams.getToDate());
 //            this.page = new AbstractMap.SimpleEntry<>("page;
 //            this.displaySize = new AbstractMap.SimpleEntry<>("displaySize;
         }
-        
+
+        List<SimpleEntry<String, ?>> getLessThanSearchTermsOnMandate() {
+            return List.of(toDate).stream().filter(e -> e.getValue() != null).collect(Collectors.toList());
+        }
+
+        List<SimpleEntry<String, ?>> getGreaterThanSearchTermsOnMandate() {
+            return List.of(fromDate).stream().filter(e -> e.getValue() != null).collect(Collectors.toList());
+        }
+
+        List<SimpleEntry<String, ?>> getExactSearchTermsOnMandate() {
+            return List.of(mandateState).stream().filter(e -> isNotBlank(e.getValue())).collect(Collectors.toList());
+        }
+
```

Comment:
I know you’re just modifying what’s already there but I find this class quite hard to follow. I’m not saying change it now, but fundamentally I don’t see why it can’t be simplified to something like this:

```java
public List<Mandate> search(MandateSearchParams mandateSearchParams) {
	var sql = new StringBuilder();
	sql.append("SELECT DISTINCT … WHERE g.external_id = :gatewayAccountExternalId");

	var sqlParams = new HashMap<>();
	sqlParams.put("gatewayAccountExternalId", mandateSearchParams.getGatewayAccountExternalId());

	Optional.ofNullable(mandateSearchParams.getReference()).ifPresent(reference -> {
		sql.append(" AND m.reference ILIKE :reference");
		sqlParams.put("reference", reference);
	});

	Optional.ofNullable(mandateSearchParams.getFromDate()).ifPresent(fromDate -> {
		sql.append(" AND m.created_date_from > :created_date_from");
		sqlParams.put("created_date_from", fromDate);
	});

	// etc. etc.

    return jdbi.withHandle(handle -> {
        Query query = handle.createQuery(sql.toString());
        sqlParams.forEach(query::bind);
        return query.map(new MandateMapper()).list();
    });
}
```

It would be more verbose but I think it would be a lot easier to follow.

Reply:
It was the verbosity I was trying to avoid! Hence my attempt to make this more elegant by returning lists and mapping over them. However I think you're right, it's getting quite hard to read and I will use the more verbose style you have suggested for readability.